File: plugins/skill-intake/scripts/prepare_notion_assets.js
#!/usr/bin/env node
'use strict';

const fs = require('node:fs');
const path = require('node:path');
const crypto = require('node:crypto');

const ALLOWED = new Set(['.png', '.svg', '.jpg', '.jpeg', '.gif']);

function walk(dir, acc = []) {
  if (!fs.existsSync(dir)) return acc;
  for (const ent of fs.readdirSync(dir, { withFileTypes: true })) {
    const p = path.join(dir, ent.name);
    if (ent.isDirectory()) walk(p, acc);
    else acc.push(p);
  }
  return acc;
}

function hash(filePath) {
  const buf = fs.readFileSync(filePath);
  return crypto.createHash('sha256').update(buf).digest('hex').slice(0, 16);
}

function prepare(srcDir, destDir) {
  fs.mkdirSync(destDir, { recursive: true });
  const items = [];
  const skipped = [];
  for (const file of walk(srcDir)) {
    const ext = path.extname(file).toLowerCase();
    if (!ALLOWED.has(ext)) { skipped.push({ path: file, reason: 'ext' }); continue; }
    const rel = path.relative(srcDir, file);
    const dest = path.join(destDir, rel);
    fs.mkdirSync(path.dirname(dest), { recursive: true });
    fs.copyFileSync(file, dest);
    const stat = fs.statSync(dest);
    items.push({
      path: rel,
      absolute: dest,
      size: stat.size,
      sha256_16: hash(dest),
      ext,
    });
  }
  const summary = {
    total: items.length,
    by_ext: items.reduce((a, it) => { a[it.ext] = (a[it.ext] || 0) + 1; return a; }, {}),
  };
  return { items, skipped, summary, src: srcDir, dest: destDir };
}

function main(argv) {
  const src = argv[2];
  const dest = argv[3];
  const out = argv[4];
  if (!src || !dest) { process.stderr.write('usage: prepare_notion_assets.js <src> <dest> [manifest.json]\n'); return 2; }
  const manifest = prepare(path.resolve(src), path.resolve(dest));
  const text = JSON.stringify(manifest, null, 2);
  const manifestPath = out ? path.resolve(out) : path.join(path.resolve(dest), 'manifest.json');
  fs.writeFileSync(manifestPath, text + '\n');
  process.stdout.write(JSON.stringify({ ok: true, manifest: manifestPath, total: manifest.summary.total }) + '\n');
  return 0;
}

module.exports = { prepare };

if (require.main === module) {
  process.exit(main(process.argv));
}

File: plugins/skill-intake/scripts/prepare_notion_assets.py
#!/usr/bin/env python3
"""Copy visual assets into a Notion-upload staging dir and emit a manifest."""
from __future__ import annotations

import hashlib
import json
import shutil
import sys
from pathlib import Path
from typing import Any

ALLOWED = {'.png', '.svg', '.jpg', '.jpeg', '.gif'}


def walk(dir_: Path) -> list[Path]:
    if not dir_.exists():
        return []
    out: list[Path] = []
    for p in dir_.rglob('*'):
        if p.is_file():
            out.append(p)
    return out


def file_hash(path: Path) -> str:
    h = hashlib.sha256()
    h.update(path.read_bytes())
    return h.hexdigest()[:16]


def prepare(src_dir: Path, dest_dir: Path) -> dict[str, Any]:
    dest_dir.mkdir(parents=True, exist_ok=True)
    items: list[dict[str, Any]] = []
    skipped: list[dict[str, str]] = []
    for file in walk(src_dir):
        ext = file.suffix.lower()
        if ext not in ALLOWED:
            skipped.append({'path': str(file), 'reason': 'ext'})
            continue
        rel = file.relative_to(src_dir)
        dest = dest_dir / rel
        dest.parent.mkdir(parents=True, exist_ok=True)
        shutil.copyfile(file, dest)
        size = dest.stat().st_size
        items.append({
            'path': str(rel),
            'absolute': str(dest),
            'size': size,
            'sha256_16': file_hash(dest),
            'ext': ext,
        })
    by_ext: dict[str, int] = {}
    for it in items:
        by_ext[it['ext']] = by_ext.get(it['ext'], 0) + 1
    return {
        'items': items,
        'skipped': skipped,
        'summary': {'total': len(items), 'by_ext': by_ext},
        'src': str(src_dir),
        'dest': str(dest_dir),
    }


def main(argv: list[str]) -> int:
    if len(argv) < 2:
        sys.stderr.write('usage: prepare_notion_assets.py <src> <dest> [manifest.json]\n')
        return 2
    src = Path(argv[0]).resolve()
    dest = Path(argv[1]).resolve()
    out = argv[2] if len(argv) > 2 else None
    manifest = prepare(src, dest)
    text = json.dumps(manifest, ensure_ascii=False, indent=2)
    manifest_path = Path(out).resolve() if out else dest / 'manifest.json'
    manifest_path.write_text(text + '\n', encoding='utf-8')
    sys.stdout.write(json.dumps({'ok': True, 'manifest': str(manifest_path), 'total': manifest['summary']['total']}, ensure_ascii=False) + '\n')
    return 0


if __name__ == '__main__':
    sys.exit(main(sys.argv[1:]))
